File: flymail/backend/modules/system/setting/baseurl_test.go
package setting

import "testing"

// 对外访问地址的校验。
//
// ── 为什么这项要认真校验 ─────────────────────────────────────────────────────
//
// 它的唯一用途是拼通知里的「打开邮件」链接。填错不会有任何即时反馈——设置页保存
// 成功、通知照发，直到某天有人点了那条链接才发现是死链。所以能在保存那一刻
// 判出来的错，必须当场拦下。
//
// 最容易漏的是 `mail.example.com` 这种没有 scheme 的写法：它看起来完全正常，
// Go 的 url.Parse 也**不会报错**（当成相对路径，Host 为空），拼出来是
// `mail.example.com/?message=1`。
func TestNormalizeBaseURL(t *testing.T) {
	ok := []struct{ in, want string }{
		{"", ""},
		{"https://mail.example.com", "https://mail.example.com"},
		// 结尾斜杠统一去掉，拼链接的地方就不必两边都判断
		{"https://mail.example.com/", "https://mail.example.com"},
		{"http://192.168.5.11:8086", "http://192.168.5.11:8086"},
		{"http://192.168.5.11:8086/", "http://192.168.5.11:8086"},
		// 反代挂在子路径下是常见部署形态
		{"https://example.com/mail", "https://example.com/mail"},
		{"  https://mail.example.com  ", "https://mail.example.com"},
		// 查询串与锚点会和我们自己的参数打架，丢掉
		{"https://mail.example.com/?x=1#top", "https://mail.example.com"},
	}
	for _, tc := range ok {
		got, err := NormalizeBaseURL(tc.in)
		if err != nil {
			t.Errorf("%q 应当合法，却报错：%v", tc.in, err)
			continue
		}
		if got != tc.want {
			t.Errorf("%q 规整成了 %q，想要 %q", tc.in, got, tc.want)
		}
	}

	bad := []struct{ in, why string }{
		{"mail.example.com", "没有 scheme：url.Parse 不报错，拼出来是死链"},
		{"//mail.example.com", "只有 // 也没有 scheme"},
		{"ftp://mail.example.com", "不是 http/https"},
		{"mailto:a@b.com", "不是 http/https"},
		{"http://", "缺少主机名"},
		{"https:///path", "缺少主机名"},
	}
	for _, tc := range bad {
		if got, err := NormalizeBaseURL(tc.in); err == nil {
			t.Errorf("%q 应当被拒（%s），却通过了，结果 %q", tc.in, tc.why, got)
		}
	}
}

// 规整结果再规整一次不变。设置页每次保存都会跑一遍，不幂等的话地址会被反复削短。
func TestNormalizeBaseURLIsIdempotent(t *testing.T) {
	for _, in := range []string{"https://mail.example.com/", "https://example.com/mail/", ""} {
		once, err := NormalizeBaseURL(in)
		if err != nil {
			t.Fatalf("%q: %v", in, err)
		}
		twice, err := NormalizeBaseURL(once)
		if err != nil {
			t.Fatalf("%q 第二次: %v", once, err)
		}
		if once != twice {
			t.Errorf("不幂等：%q → %q → %q", in, once, twice)
		}
	}
}

File: flymail/frontend/src/lib/base-url.test.ts
import { describe, it, expect } from 'vitest'
import { checkBaseUrl } from '@/lib/base-url'

/**
 * 「对外访问地址」的校验。
 *
 * ── 为什么这项要认真校验 ─────────────────────────────────────────────────────
 *
 * 它唯一的用途是拼通知里的「打开邮件」链接。填错没有任何即时反馈——设置保存成功、
 * 通知照发，直到某天有人点了那条链接才发现打不开。能在保存那一刻判出来的错，
 * 必须当场拦下。
 *
 * 最容易漏的是 `mail.example.com` 这种没有 scheme 的写法：看起来完全正常，
 * 后端的 url.Parse 也不会报错（当成相对路径），拼出来是死链。
 */
describe('checkBaseUrl', () => {
  it('留空是合法的——表示不配，通知就不带链接', () => {
    expect(checkBaseUrl('')).toBe('ok')
    expect(checkBaseUrl('   ')).toBe('ok')
  })

  it('接受带主机名的 http/https 绝对地址', () => {
    for (const v of [
      'https://mail.example.com',
      'https://mail.example.com/',
      'http://192.168.5.11:8086',
      'https://example.com/mail', // 反代挂在子路径下
      'HTTPS://MAIL.EXAMPLE.COM', // scheme 大小写不敏感
    ]) {
      expect(checkBaseUrl(v), `${v} 应当合法`).toBe('ok')
    }
  })

  it('没有 scheme 的裸主机名要拒掉', () => {
    // ⚠ 这条是重点：它看起来最像"对的"，拼出来却是 mail.example.com/?message=1
    expect(checkBaseUrl('mail.example.com')).toBe('scheme')
    // 192 不是合法 scheme（必须字母开头），所以这也落进"没有 scheme"
    expect(checkBaseUrl('192.168.5.11:8086')).toBe('scheme')
  })

  it('非 http/https 的 scheme 要拒掉', () => {
    expect(checkBaseUrl('ftp://mail.example.com')).toBe('scheme')
    expect(checkBaseUrl('mailto:a@b.com')).toBe('scheme')
  })

  /**
   * ⚠ 这几条钉的是「前端判据不能比后端松」。
   *
   * `new URL()` 是 WHATWG 解析器，对 http/https 会自动补斜杠，于是
   * `http:/mail.example.com`（少打一个斜杠）在它眼里完全合法、hostname 还是对的；
   * 而后端 Go 的 url.Parse 认为 Host 为空，直接拒。只依赖 URL 构造函数的话，
   * 前端放行 → 用户点保存 → 撞一个 400。
   *
   * 判据往松了写在这里是**静默**的：功能看起来都正常，只有真去点保存才暴露。
   */
  it('有 scheme 但缺 //主机名 的要拒掉，判据不能比后端松', () => {
    expect(checkBaseUrl('http://'), 'http:// 后面空着').toBe('host')
    expect(checkBaseUrl('https:///path'), '三个斜杠，主机名位置是空的').toBe('host')
    expect(checkBaseUrl('http:/mail.example.com'), '少打一个斜杠——最常见的手误').toBe('host')
    expect(checkBaseUrl('http:mail.example.com'), '冒号后面直接跟主机名').toBe('host')
  })

  it('完全不是 URL 的输入不抛异常', () => {
    // 没有冒号 → 连 scheme 都没有，提示"必须以 http:// 开头"正是用户要看的
    expect(checkBaseUrl('这是一段说明文字')).toBe('scheme')
    // 形式对但 URL 构造函数仍解析不了的，落到 invalid
    expect(checkBaseUrl('http://[')).toBe('invalid')
  })
})
